examples/app/headless_renderer.rs

```diff
@@ -0,0 +1,547 @@
+                    sample_count: 1,
+                    usage: TextureUsages::COPY_DST | TextureUsages::TEXTURE_BINDING,
+                    view_formats: &[],
+                },
+                ..Default::default()
+            };
+            cpu_image.resize(size);
+            let cpu_image_handle = images.add(cpu_image);
+
+            commands.spawn(ImageCopier::new(
+                render_target_image_handle.clone(),
+                size,
+                render_device,
+            ));
+
+            commands.spawn(ImageToSave(cpu_image_handle));
+
+            scene_controller.state = SceneState::Render(pre_roll_frames);
+            scene_controller.name = scene_name;
+            RenderTarget::Image(render_target_image_handle)
+        }
+
+        // Takes from channel image content sent from render world and saves it to disk
+        fn update(
+            images_to_save: Query<&ImageToSave>,
+            receiver: Res<MainWorldReceiver>,
+            mut images: ResMut<Assets<Image>>,
+            mut scene_controller: ResMut<SceneController>,
+            mut app_exit_writer: EventWriter<AppExit>,
+        ) {
+            if let SceneState::Render(n) = scene_controller.state {
+                if n < 1 {
+                    // We don't want to block the main world on this,
+                    // so we use try_recv which attempts to receive without blocking
+                    let mut image_data = Vec::new();
+                    while let Ok(data) = receiver.try_recv() {
+                        // image generation could be faster than saving to fs,
+                        // that's why use only last of them
+                        image_data = data;
+                    }
+                    if !image_data.is_empty() {
+                        for image in images_to_save.iter() {
+                            let img_bytes = images.get_mut(image.id()).unwrap();
+                            img_bytes.data = image_data.clone();
+
+                            let img = match img_bytes.clone().try_into_dynamic() {
+                                Ok(img) => img.to_rgba8(),
+                                Err(e) => panic!("Failed to create image buffer {e:?}"),
+                            };
+
+                            let images_dir =
+                                PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test_images");
+                            info!("Saving image to: {images_dir:?}");
+                            std::fs::create_dir_all(&images_dir).unwrap();
+
+                            let mut number = 0;
+                            let mut image_path = images_dir.join(format!("{number:03}.png"));
+                            while image_path.exists() {
+                                number += 1;
+                                image_path = images_dir.join(format!("{number:03}.png"));
+                            }
+                            if let Err(e) = img.save(image_path) {
```

Comment:
I think this is a bad pattern to show in example, as it write to the file system in a blocking way, but moving this to a task will add more complexity. it needs at least a comment

Reply:
Comments have been added